ui/dashboards/src/utils/time-range-params.ts

```diff
@@ -0,0 +1,111 @@
+// Copyright 2022 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { useMemo, useEffect, useRef } from 'react';
+import { useQueryParams, DecodedValueMap, StringParam } from 'use-query-params';
+import { getUnixTime, sub } from 'date-fns';
+import {
+  TimeRangeValue,
+  AbsoluteTimeRange,
+  toAbsoluteTimeRange,
+  isRelativeTimeRange,
+  isDurationString,
+  RelativeTimeRange,
+  DurationString,
+  isAbsoluteTimeRange,
+  parseDurationString,
+} from '@perses-dev/core';
+import { useSelectedTimeRange } from '../context';
+
+const timeRangeQueryConfig = {
+  start: StringParam,
+  end: StringParam,
+};
+
+/**
+ * Gets the initial time range taking into account URL params and dashboard JSON duration
+ * Sets start query param if it is empty on page load
+ */
+export function useInitialTimeRange(dashboardDuration: DurationString): TimeRangeValue {
+  const [query, setQuery] = useQueryParams(timeRangeQueryConfig);
+  const { start, end } = query;
+
+  return useMemo(() => {
+    let initialTimeRange: TimeRangeValue = { pastDuration: dashboardDuration };
+    if (!start) {
+      setQuery({ start: dashboardDuration });
+      return initialTimeRange;
+    }
+
+    if (isDurationString(start.toString())) {
+      initialTimeRange = { pastDuration: start } as RelativeTimeRange;
+    } else {
+      initialTimeRange = {
+        start: new Date(Number(start)),
+        end: end ? new Date(Number(end)) : new Date(),
+      };
+    }
+    return initialTimeRange;
+  }, [start, end, dashboardDuration, setQuery]);
+}
+
+/**
+ * Set start and end URL params and update when selected time range changes
+ */
+export function useSyncTimeRangeParams() {
+  const { selectedTimeRange } = useSelectedTimeRange();
+  const [query, setQuery] = useQueryParams(timeRangeQueryConfig);
+  const lastParamSync = useRef<DecodedValueMap<typeof timeRangeQueryConfig>>();
```

Comment:
I'm not sure if I'm following this part. Why do we need to keep track of this?

Reply:
There may be a better / less complicated way to handle this, but currently it's needed for the back button to function (see the two `lastStart` conditions: [1](https://github.com/perses/perses/blob/33a590cee37a10f1cee92d98348df6417ea268e7/ui/dashboards/src/utils/time-range-params.ts#L74-L77), [2](https://github.com/perses/perses/blob/33a590cee37a10f1cee92d98348df6417ea268e7/ui/dashboards/src/utils/time-range-params.ts#L83-L88) and then useSyncActiveTimeRange). 

https://github.com/perses/perses/blob/33a590cee37a10f1cee92d98348df6417ea268e7/ui/dashboards/src/utils/time-range-params.ts#L74-L77

https://github.com/perses/perses/blob/33a590cee37a10f1cee92d98348df6417ea268e7/ui/dashboards/src/utils/time-range-params.ts#L83-L88

Reply:
Are we talking about the browser back button? If so, does this work if I (for example) use the back button to go back two (or more) history entries rather than just one in a single click (i.e. click and hold the back button, select an entry farther back in the stack)?

Feels like we should try to figure out what's up and try to make it less complicated if we can. Maybe there's some state-syncing going on somewhere that we can refactor to avoid or something.

Reply:
So it did work when pressing back multiple times due to the logic in useSyncActiveTimeRange, but agreed this approach was extremely hacky. I updated this PR based on our discussion yesterday so useSyncTimeRangeParams is gone[ in favor of useSetTimeRangeParams ](https://github.com/perses/perses/pull/621/files#diff-6f7f1406b070f8ac3958be3136acc07bf54dc7f7725013108f4813ac8583b903)